[2.9] Select default charm url depending on controller version

achilleasa opened the pull request:
Prior to this PR, 2.9 clients always assumed charmhub (ch) to be the
default charm URL schema for charms URLs without one. This breaks
backward compatibility with older controllers when a 2.9 client attempts
to deploy a charm (`juju deploy wordpress`) without an explicit schema
prefix to a 2.8 controller.

In this case, the client assumes that we try to deploy a charmhub charm to the
controller but since the controller lacks the appropriate APIs, the
charm API client returns back a NotSupported error which the charm
adapter spits out to the user with a totally incomprehensible error:
"ERROR resolve charms not supported".

To address this problem, the client will first check the supported charm
API version and use this information to select the default schema for
schema-less charm URLs. This is accomplished via the updated
charm/v8.EnsureSchema signature which requires the default schema to be
specified.

The default schema is threaded through the various deploy-related
structs (charms/bundles etc.) and is also used by the `juju refresh` command.

Finally, if the user tries to explicitly deploy a charmhub charm against
a pre 2.9 controller, they will get a straightforward error with a
recommendation to update to 2.9+ if they want to use charmhub.

## QA steps

```sh
$ sudo snap install --channel 2.8 juju
$ /snap/bin/juju bootstrap lxd test

# Using the 2.9 client from this PR
$ juju deploy ceph-mon
Located charm "ceph-mon" in charm-store, revision 55
Deploying "ceph-mon" from charm-store charm "ceph-mon", revision 55 in channel stable

$ juju refresh ceph-mon
Looking up metadata for charm-store charm "ceph-mon"
ERROR already running latest charm "ceph-mon"

$ juju refresh ceph-mon --switch influxdb
Looking up metadata for charm-store charm "ceph-mon"
Added charm-store charm "influxdb", revision 23 in channel stable, to the model
Adding endpoint "grafana-source" to default space "alpha"
Adding endpoint "query" to default space "alpha"
Leaving endpoints in "alpha": admin, nrpe-external-master

$ juju refresh ceph-mon --switch ch:wordpress                                                                                               
Looking up metadata for charm-store charm "influxdb"
ERROR charmhub charms are not supported by the current controller; if you wish to use charmhub consider upgrading your controller to 2.9+.

$ juju deploy ch:wordpress
ERROR charmhub charms are not supported by the current controller; if you wish to use charmhub consider upgrading your controller to 2.9+.

# Use the following bundle
$ cat b.yaml
series: bionic
applications:
  ceph-mon:
    charm: ceph-mon
    num_units: 1

$ juju remove-application ceph-mon
$ juju deploy ./b.yaml
Located charm "ceph-mon" in charm-store, channel stable
Executing changes:
- upload charm ceph-mon from charm-store for series bionic
- deploy application ceph-mon from charm-store on bionic
- add unit ceph-mon/7 to new machine 7
Deploy of bundle completed.

# Now try this bundle
$ cat ch.yaml
series: bionic
applications:
  wordpress:
    charm: ch:wordpress
    num_units: 1

$ juju deploy ./ch.yaml
ERROR cannot deploy bundle: cannot resolve charm or bundle "wordpress": charmhub charms are not supported by the current controller; if you wish to use charmhub consider upgrading your controller to 2.9+
```

Please also try the same steps against a 2.9 controller.

## Bug reference
https://bugs.launchpad.net/juju/+bug/1927091

hmlanigan commented on cmd/juju/application/refresh.go:
```diff
@@ -502,7 +502,7 @@ func (c *refreshCommand) isCharmHubWithRevision(source commoncharm.OriginSource)
 	}
 	// EnsureSchema will error if input is an empty string.
-	path, err := charm.EnsureSchema(c.SwitchURL)
+	path, err := charm.EnsureSchema(c.SwitchURL, charm.CharmHub)
```
Why are we assuming the switch charm must be charmhub if not specified, with a 2.8 controller and 2.9 client?

$ juju refresh ceph-mon --switch influxdb
Looking up metadata for charm-store charm "ceph-mon"
ERROR charmhub charms are not supported by the current controller; if you wish to use charmhub consider upgrading your controller to 2.9+.
$ juju refresh ceph-mon --switch cs:influxdb
Looking up metadata for charm-store charm "ceph-mon"
Added charm-store charm "influxdb", revision 23 in channel stable, to the model
Adding endpoint "grafana-source" to default space "alpha"
Adding endpoint "query" to default space "alpha"
Leaving endpoints in "alpha": admin, nrpe-external-master

achilleasa replied:
We shouldn't. I was under the wrong impression that this codepath was only used for 2.9+. I will update it to use the same logic as the deployer.

achilleasa replied:
Pushed https://github.com/juju/juju/pull/12948/commits/1ce13dfedb0124b24550cc2bd865bf61fa26be17 to address these comments and updated QA steps.

hmlanigan commented on cmd/juju/application/refresher/refresher.go:
```diff
@@ -303,7 +303,7 @@ type charmStoreRefresher struct {
 // Depending on the charm url, will then determine if that's true or not.
 func (r *charmStoreRefresher) Allowed(cfg RefresherConfig) (bool, error) {
-	path, err := charm.EnsureSchema(cfg.CharmRef)
+	path, err := charm.EnsureSchema(cfg.CharmRef, charm.CharmHub)
```
should be charm.CharmStore for the charmStoreRefresher

achilleasa replied:
Nice catch!

hmlanigan reviewed (changes requested):
deploy looks good.  tested with 2.8 and 2.9 controller.

questions on refresh though.

achilleasa commented on cmd/juju/application/store/charmadapter.go:
```diff
@@ -75,7 +75,7 @@ func (c *CharmAdaptor) ResolveCharm(url *charm.URL, preferredOrigin commoncharm.
 	if errors.IsNotSupported(err) {
 		if charm.CharmHub.Matches(url.Schema) {
-			return nil, commoncharm.Origin{}, nil, errors.Trace(err)
+			return nil, commoncharm.Origin{}, nil, errors.NewNotSupported(nil, "charmhub charms are not supported by the current controller; if you wish to use charmhub consider upgrading your controller to 2.9+.")
```
`errors.NotSupportedf` injects a "not supported" suffix to whatever message you give it. In this case, I wanted a NotSupported error with a custom message hence the use of the alternative constructor.

achilleasa commented:
@hmlanigan With my latest commit, refresh --switch works. But now I hit another issue
(using the client from this PR and a 2.8/stable controller):

```
$ juju status
Model  Controller  Cloud/Region            Version  SLA          Timestamp
test   test        lxd-with-cache/default  2.8.10   unsupported  11:57:05+01:00


$ juju deploy cs:ceph-mon
$ juju refresh ceph-mon --switch cs:influxdb
Looking up metadata for charm-store charm "ceph-mon"
Added charm-store charm "influxdb", revision 23 in channel stable, to the model
Adding endpoint "grafana-source" to default space "alpha"
Adding endpoint "query" to default space "alpha"
Leaving endpoints in "alpha": admin, nrpe-external-master

$ juju refresh ceph-mon --switch cs:ceph-mon
Looking up metadata for charm-store charm "influxdb"
Added charm-store charm "ceph-mon", revision 55 in channel stable, to the model
ERROR cannot upgrade application "ceph-mon" to charm "cs:ceph-mon-55": state changing too quickly; try again soon (excessive contention)
```

hmlanigan reviewed (approved):
QA is good.

achilleasa commented:
$$merge$$
